### PR TITLE
Handling configuration exceptions in IntegrationRunner

### PR DESCRIPTION
**Pull request by akashkulk**
Breaking off this PR: https://github.com/airbytehq/airbyte/pull/18560
into smaller pieces.

1. `IntegrationRunner.java` : All config errors are handled here, rather than in the individual protocol methods (read(), check() etc). A trace message is emitted for every config error.
2. `ConfigErrorException.java` : New exception class `ConfigErrorException` indicates an issue with the user's configuration.

This should unblock other use cases of throwing a `ConfigErrorException`. Remaining work is still required to: 
- Remove catching/rethrowing exceptions in each of the sub connector classes. 
- Migrating existing config errors (cursor exceptions, SSH exceptions, connection errors) to be a config error

### CODE REVIEW

**Review by rodireich — APPROVED** [review]

File: airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/base/IntegrationRunner.java [hunk 1]
```diff
@@ -196,11 +233,11 @@ private static void runConsumer(final AirbyteMessageConsumer consumer) throws Ex
    */
   @VisibleForTesting
   static void watchForOrphanThreads(final Procedure runMethod,
-                                    final Runnable exitHook,
```
Comment (ryankfu) [inline airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/base/IntegrationRunner.java:199]: these tab/space amounts probably shouldn't be changed to match how other method's space out their arguments. Another benefit in this case is that it's easier to see that this method throws an Exception

File: airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/base/IntegrationRunner.java [hunk 2]
```diff
@@ -105,53 +108,87 @@ private void runInternal(final IntegrationConfig parsed) throws Exception {
...
+      // If the source connector throws a config error, a trace message with the relevant message should
+      // be surfaced.
+      if (isConfigError(e)) {
+        AirbyteTraceMessageUtility.emitConfigErrorTrace(e, displayMessage);
+      }
+      if (parsed.getCommand().equals(Command.CHECK)) {
+        // Currently, special handling is required for the SPEC case since the user display information in
+        // the trace message is
+        // not properly surfaced to the FE. In the future, we can remove this and just throw an exception.
+        outputRecordCollector.accept(new AirbyteMessage().withType(Type.CONNECTION_STATUS).withConnectionStatus(
```
Comment (ryankfu) [inline airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/base/IntegrationRunner.java:166]: same comment as before, can this be spaced out for readability? 

https://github.com/airbytehq/airbyte/blob/9c2601013771ca0cd83b633303d40dcbc1a38b1a/airbyte-config/config-persistence/src/test/java/io/airbyte/config/persistence/StatePersistenceTest.java#L104-L111

**Review by ryankfu — APPROVED** [review]
nothing blocking, just minor comments on keeping things either the same or spacing out for readability

**Review by tuliren — COMMENTED** [review]
Some of the format changes can be auto fixed by running `./gradlew :spotlessJavaApply`. This command will format all Java code. You can then check in format updates relevant to your changes.

File: airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/base/IntegrationRunner.java (hunk 2 above)
Comment (akashkulk) [reply airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/base/IntegrationRunner.java:166]: Thanks, I ran the formatting command on these files, so I'll defer to that

File: airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/base/IntegrationRunner.java (hunk 1 above)
Comment (akashkulk) [reply airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/base/IntegrationRunner.java:199]: Thanks, I ran the formatting command on these files, so I'll defer to that

**Pull request merged**